src/persist-client/src/operators/shard_source.rs

```diff
@@ -64,7 +134,7 @@ use crate::{PersistLocation, ShardId};
 /// using [`timely::dataflow::operators::generic::operator::empty`].
 ///
 /// [advanced by]: differential_dataflow::lattice::Lattice::advance_by
-pub fn shard_source<K, V, D, G>(
+pub fn shard_source<K, V, E, D, G>(
```

Comment:
I don't love needing a new parameter or the emit_fn here, but I was unable to split decoding from MFP and [pass the feature benchmark](https://buildkite.com/materialize/nightlies/builds/1882)... in particular the feature benchmark has many `count(*)` queries that MFP and best-effort consolidation is extremely effective at optimizing.